processor/src/main/java/org/mapstruct/ap/internal/util/Strings.java

```diff
@@ -183,7 +183,15 @@ public static String sanitizeIdentifierName(String identifier) {
 
             if ( firstAlphabeticIndex < identifier.length()) {
                 // If it is not consisted of only underscores
-                return identifier.substring( firstAlphabeticIndex ).replace( "[]", "Array" );
+                String firstAlphaString = identifier.substring( firstAlphabeticIndex ).replace( "[]", "Array" );
+
+                return firstAlphaString.codePoints()
+                    .map(codePoint -> Character.isJavaIdentifierPart(codePoint) || codePoint == (int) '.' ? codePoint : (int) '_')
+                    .collect(
+                        StringBuilder::new,
+                        StringBuilder::appendCodePoint,
+                        StringBuilder::append
+                    ).toString();
```

Comment:
`codePoints()` is available from Java 9, we can't use this in this path.

In addition to that `codePoint` can never be `.`, since `.` are treated in a special way in MapStruct for nesting mappings.

Let's use good old iteration here.